### PR TITLE
Agda: 2.4.2.3 -> 2.5.1

### LINKED ISSUE #15306: AgdaStdlib – permission denied in build
## Issue description
As of Agda 2.5.1 (which comes on the latest nixos-unstable release), both nixpkgs' agda-stdlib (from git nominally at version 2.4.2.3) and an updated version (v0.12, the most recent .tar.gz release) fail to build. Some modules are checked successfully, but it gets stuck at `Agda.Builtin.Unit` (the first builtin module it finds). For this module only, it tries to write to `/nix/store/`.

Output:
```
building path(s) ‘/nix/store/p8nqivxn3zr6mmwis8xgyshjgqhdxgf8-agda-stdlib-v0.12’
unpacking sources
unpacking source archive /nix/store/yr0h4ia34i9q6mqnslqgm1a5draksxyg-v0.12.tar.gz
source root is agda-stdlib-0.12
setting SOURCE_DATE_EPOCH to timestamp 1460834616 of file agda-stdlib-0.12/.travis.yml
patching sources
configuring
no configure script, doing nothing
building
Checking Everything (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/Everything.agda).
 Checking Algebra (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Algebra.agda).
  Checking Relation.Binary (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Relation/Binary.agda).
   Checking Data.Product (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Data/Product.agda).
    Checking Function (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Function.agda).
     Checking Level (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Level.agda).
     Finished Level.
    Finished Function.
    Checking Relation.Nullary (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Relation/Nullary.agda).
     Checking Data.Empty (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Data/Empty.agda).
     Finished Data.Empty.
    Finished Relation.Nullary.
   Finished Data.Product.
   Checking Data.Sum (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Data/Sum.agda).
    Checking Data.Maybe.Base (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Data/Maybe/Base.agda).
     Checking Data.Bool.Base (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Data/Bool/Base.agda).
      Checking Data.Unit.Base (/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Data/Unit/Base.agda).
       Checking Agda.Builtin.Unit (/nix/store/qpzrr8ymvh7ay5z96fci0qp71n4x990k-Agda-2.5.1/share/x86_64-linux-ghc-7.10.3/Agda-2.5.1/lib/prim/Agda/Builtin/Unit.agda).
Failed to write interface /nix/store/qpzrr8ymvh7ay5z96fci0qp71n4x990k-Agda-2.5.1/share/x86_64-linux-ghc-7.10.3/Agda-2.5.1/lib/prim/Agda/Builtin/Unit.agdai.
/tmp/nix-build-agda-stdlib-v0.12.drv-0/agda-stdlib-0.12/src/Data/Unit/Base.agda:13,13-30
/nix/store/qpzrr8ymvh7ay5z96fci0qp71n4x990k-Agda-2.5.1/share/x86_64-linux-ghc-7.10.3/Agda-2.5.1/lib/prim/Agda/Builtin/Unit.agdai:
openBinaryFile: permission denied (Permission denied)
builder for ‘/nix/store/z7nzphrz1m57mzp7516w53sl5nfa7a81-agda-stdlib-v0.12.drv’ failed with exit code 1
error: build of ‘/nix/store/z7nzphrz1m57mzp7516w53sl5nfa7a81-agda-stdlib-v0.12.drv’ failed
```

### Steps to reproduce
```
nix-build -A AgdaStdlib
```

## Technical details

* System: 16.09.git.39b3345 (Flounder)
* Nix version: nix-env (Nix) 1.11.2
* Nixpkgs version: "16.09.git.39b3345"

### PR DESCRIPTION
**Pull request by laMudri**
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This PR contains a few things associated with updating from `Agda-2.4.2.3` to `Agda-2.5.1`. It fixes #15306 by compiling all of the built-in modules in the `postInstall` section. Then, `agda-stdlib` and `agda-prelude` have been updated to versions which work with the new version.

I've tested `agda-mode` and tested checking modules using `agda-stdlib`. I don't know what `agda-ghc-names` is about, but I vaguely assume that it works. `agda-ghc-names --help` works, anyway.

### CODE REVIEW

**Comment by laMudri** [comment]
Okay, so the update breaks some old packages from 2014. Some of the Agda packages seem to be obsolete (`agda-base` replaced by `HoTT-Agda`, `Agda-Sheaves` replaced by `constructive-sheaf-semantics`). Should I go through them all and set `broken = true`?

**Comment by jagajaga** [comment]
cc @peti

**Comment by laMudri** [comment]
Tests are failing now because it takes too long. Agda does take a long time to build.

**Comment by joachifm** [comment]
Merging, given the lack of feedback and seeing how this fixes an issue with Agda. Thank you.

File: pkgs/development/haskell-modules/hackage-packages.nix [hunk 1]
```diff
@@ -847,13 +847,22 @@ self: {
        ];
        executableToolDepends = [ emacs ];
        postInstall = ''
```
Comment (jagajaga) [inline pkgs/development/haskell-modules/hackage-packages.nix]: @joachifm I'm pretty sure that manually editing of this file is denied.

Comment (jagajaga) [inline pkgs/development/haskell-modules/hackage-packages.nix]: And I'm right. 
> /* hackage-packages.nix is an auto-generated file -- DO NOT EDIT! */

Comment (jagajaga) [inline pkgs/development/haskell-modules/hackage-packages.nix]: Please revert and fix.

Comment (joachifm) [inline pkgs/development/haskell-modules/hackage-packages.nix]: Done.

Comment (joachifm) [inline pkgs/development/haskell-modules/hackage-packages.nix]: Well, the revert part.

**Comment by peti** [comment]
@joachifm, you diagnose "lack of feedback" in a PR that's barely 24 hours old?

**Comment by joachifm** [comment]
@peti for me it's showing 2 days. Anyway, I was too hasty, I'm sorry.

**Comment by joachifm** [comment]
@laMudri I've pushed a fixed version of your PR at https://github.com/joachifm/nixpkgs/commits/Agda-fixed

**Comment by laMudri** [comment]
Thanks for fixing, and sorry about not looking more carefully. How did the `postInstall` instructions ever get into that auto-generated file? I notice that other Agda versions also have those commands, but they're not in `configuration-common.nix`.

**Comment by peti** [comment]
https://github.com/NixOS/cabal2nix/blob/master/distribution-nixpkgs/src/Distribution/Nixpkgs/Haskell/FromCabal/PostProcess.hs#L193-L199

**Comment by laMudri** [comment]
Um, too much was reverted. Please see 072ce88, 92b843f, and 6343cf8. All of these packages are still broken on master.

**Comment by joachifm** [comment]
@laMudri I reverted the merge itself. Please open a new PR if you want these changes integrated. Feel free to use the version I posted, but beware that it is untested, so please confirm that it is still valid.

I think the only issue was writing to the auto-generated file, so if you address that we can go ahead with your changes. The issue is quite old at this point (17 days as of this writing) with no alternative solution.

**Comment by laMudri** [comment]
Okay, thanks. I have something tested based on your branch, so I'll just double check it and make a new PR.

**Pull request merged**